src/main/webapp/go.js

```diff
@@ -0,0 +1,143 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+const GO_URL = '/go-data';
+
+const INVISIBLE_CLASS = 'slide-invisible';
+const NAME_URL = '/name-data';
+const START_ID = 'start-button';
+const PROCEED_ID = 'proceed-button';
+const RIDDLE_ID = 'riddle-area';
+const INDEX_PARAM = 'new-index';
+const FINAL_MSSG = 'Congrats, you\'ve finished the hunt!';
+
+const riddleArr = [];
+let destIndex;
+const destArr = [];
+
+window.onload = getHunt();
+
+/**
+ * Retrieves scavenger hunt data, and updates to the current destination
+ * to reflect the current state of the hunt.
+ */
+function getHunt() {
+  fetch(GO_URL).then((response) => response.json()).then((mssg) => {
+    destIndex = mssg.index;
+    for (let i = 0; i < mssg.items.length; i++) {
+      riddleArr.push(mssg.items[i].riddle.puzzle);
+      destArr.push(mssg.items[i].name);
+    }
+    updateToCurrentState(destIndex);
+  });
+}
+
+/**
+ * Updates the hunt to the current destination that the user is on.
+ *
+ * @param {int} index The index of the destination that the user needs
+ * to find.
+ */
+function updateToCurrentState(index) {
+  if (index >= 0) {
+    hideStartButton();
+    changeRiddleMessage(riddleArr[index]);
+    getDestName();
+  }
+  sendIndexToServlet(index);
+  hideProceedButton();
+}
+
+/**
+ * Retrieves the string the user submitted as the destination name.
+ */
+function getDestName() {
+  fetch(NAME_URL).then((response) => response.json()).then((mssg) => {
+    if (mssg === destArr[destIndex]) { // The user entered the correct name.
+      showProceedButton();
+    }
+  });
+}
+
+/**
+ * Creates a new paragraph element from text.
+ *
+ * @param {String} text to be displayed on main page.
+ * @return {String} newLine New element to be appended to main page.
+ */
+function createLine(text) {
+  const newLine = document.createElement('p');
+  newLine.innerText = text;
+  return newLine;
+}
+
+/**
+ * Hide the proceed button.
+ */
+function hideProceedButton() {
+  const proceedButton = document.getElementById(PROCEED_ID);
+  proceedButton.classList.add(INVISIBLE_CLASS);
+}
+
+/**
+ * Show the proceed button.
+ */
+function showProceedButton() {
+  const proceedButton = document.getElementById(PROCEED_ID);
+  proceedButton.classList.remove(INVISIBLE_CLASS);
+}
```

Comment:
Suggestion: collapse these two methods into a single method that just toggles the element's visibility.

```suggestion
/**
 * Show or hide the proceed button.
 */
function toggleProceedButton(const hide) {
  const proceedButton = document.getElementById(PROCEED_ID);
  if (hide) {
    proceedButton.classList.add(INVISIBLE_CLASS);
  } else {
    proceedButton.classList.remove(INVISIBLE_CLASS);
  }
}

```